crates/ruff/src/rules/ruff/rules/unnecessary_iterable_allocation_for_first_element.rs

```diff
@@ -175,47 +214,98 @@ fn iterable_name<'a>(expr: &'a Expr, model: &SemanticModel) -> Option<&'a str> {
                 return None;
             }
 
+            let [arg] = args.as_slice() else {
+                return None;
+            };
+
             if !model.is_builtin(id.as_str()) {
                 return None;
             }
 
-            match args.first() {
-                Some(Expr::Name(ast::ExprName { id: arg_name, .. })) => Some(arg_name.as_str()),
-                Some(Expr::GeneratorExp(ast::ExprGeneratorExp {
+            match arg {
+                Expr::GeneratorExp(ast::ExprGeneratorExp {
+                    elt, generators, ..
+                }) => match_simple_comprehension(elt, generators)
+                    .map(|range| IterationTarget {
+                        range,
+                        iterable: false,
+                    })
+                    .or_else(|| {
+                        Some(IterationTarget {
+                            range: arg.range(),
+                            iterable: true,
+                        })
+                    }),
```

Comment:
Nit: I personally prefer match statements when handling both the `Some` and `None` case (It's likely that our clippy pedantic complains that you should use `if... else`. I strongly disagree with that because I find match easier to read but that works too :wink: 

```suggestion
                }) => match match_simple_comprehension(elt, generators) {
                	Some(range) =>  IterationTarget { range, iterable: false },
                	None => IterationTarget { range: arg.range(), iterable: true },
              	},
```